### ISSUE TITLE
Enable policy writers to define and reuse composite functions in order to simplify Apply expressions across policies

### ISSUE THREAD

**Issue by cdanger** [oasis-tcs/xacml-spec#34]
#22 and #23 showed the need to add new XACML functions in order to simplify very common Apply expressions that _compose_ (math sense) two or more functions together.  

For example, in #23 , the proposed `empty-bag` (resp. `non-empty-bag`) function is a composition of `bag-size` and` integer-equal `(resp. `integer-greater-than`) functions.

So instead of adding such functions to the XACML spec every time we need a new one, we could let policy writers define their own composite functions in XACML at a global level (like #12 ?) so that they can reuse them in Apply expressions, furthermore share them with other policy writers, and even further with the whole XACML community, on the TC's github for instance, as a community extension to the standard.

TODO: 

1. Define a XML schema for a composite `<FunctionDefinition>`
2. Give a few examples based on #22 and #23 
3. Define a standard package format or parent XML element for sharing such `<FunctionDefinition>`s

TO BE COMPLETED

**Comment by steven-legg** [comment]
What you are talking about is equivalent to a global/shared variable (#12) with parameters. A  function with no parameters is the same as a variable definition. A common mechanism could be defined.

**Comment by steven-legg** [comment]
I can think of a roundabout way to calculate minimum or maximum with existing functions but not a way to calculate sum or average.

**Comment by cdanger** [comment]
> I can think of a roundabout way to calculate minimum or maximum with existing functions but not a way to calculate sum or average.

Indeed, I don't see how to do sum either. But having sum defined in the standard is enough to express the average based on the sum (divided by size) in a function definition, as I did in the example above.

**Comment by steven-legg** [comment]
> Yes, sure. GlobalVariableDefinitions and FunctionDefinitions can be put in the same package.

The synergy disappears if you are going to exclude attribute designators and selectors from function definitions. My PAPs support something that looks like global variables and I have a non-trivial policy collection where every single global variable definition depends on at least one attribute designator. The usefulness of parameterized global variables would be severely limited if they couldn't also use attribute designators.

The roundabout way I thought of to calculate minimum or maximum depends on the Select expression from the Entities profile, which isn't part of your function definition. However, Select and the other quantified expressions have embedded Expression elements and so would need new rewrites for use in function definitions in order to exclude designators and selectors.

In any case, I can bypass the prohibition on attribute designator and selector elements by using the attribute-designator and attribute-selector functions from the Entities profile.

**Comment by steven-legg** [comment]
@cdanger What is the rationale for excluding attribute designators and selectors? Is it because they would be hidden inputs to the function? Perhaps a different name is in order, e.g., method, because designators would be very useful. A policy writer can always choose not to use designators.

**Comment by cdanger** [comment]
Because I have this hard habit to use the term "function" only for _pure functions_, which have nice properties that functional programmers would expect (output depends only on input, no side effects). As I had the impression that all functions in the XACML core spec were pure functions, I assumed this was implicit. But checking again... I was wrong :-( . At least `access-permitted` (which I don't use) is impure, and so are the attribute-designator and attribute-selector (new to me) that you mentioned earlier. **Sorry for the confusion.** Maybe I should rename it to PureFunctionDefinition... 

Anyway, if you are just looking for reusable expressions with attribute designators and selectors, it seems to me you can do it with GlobalVariableDefinitions already, and you don't need all this.

**Comment by steven-legg** [comment]
> Anyway, if you are just looking for reusable expressions with attribute designators and selectors, it seems to me you can do it with GlobalVariableDefinitions already, and you don't need all this.

The current global variable proposal doesn't include parameterization, but it would be a useful addition. Parameterized global variable definitions would be a superset of pure, composite functions making pure, composite functions redundant complexity.

**Comment by cdanger** [comment]
Not exactly the same though, the guarantees are much different (due to the constraint of _not_ depending on attributes / external variables - only input ones - internally or any external state). If I want to call a function / expression with the guarantees of determinism / predictability, no side effect, referential transparency (I can cache the result), immutability... then I want a pure function, not any variable expression. Also pure functions can be composed together (guarantees preserved), i.e. a composition of pure functions _is_ a pure function.

**Comment by cdanger** [comment]
Seems not to be a strong need for this feature, and there is an overlap with the GlobalVariableDefinitions in #12. So putting it on low priority for now.

**Comment by cdanger** [comment]
I don't have an immediate use case that could not be addressed by (parameterizable) SharedVariableDefinitions for now, so closing this. Happy to re-open it if the need arises again.
